image/mkosi/secure-boot/generate_nvram_vars.sh

```diff
@@ -0,0 +1,107 @@
+        return
+    fi
+    if ! test -f "${image_path}"; then
+        echo "Image \"${image_path}\" does not exist yet. To generate nvram, create disk image first."
+        return
+    fi
+
+    OVMF_CODE=/usr/share/OVMF/OVMF_CODE_4M.ms.fd
+    OVMF_VARS=/usr/share/OVMF/OVMF_VARS_4M.ms.fd
+    if ! test -f "${OVMF_CODE}"; then
+        OVMF_CODE=/usr/share/OVMF/OVMF_CODE.secboot.fd
+    fi
+    if ! test -f "${OVMF_VARS}"; then
+        OVMF_VARS=/usr/share/OVMF/OVMF_VARS.secboot.fd
+    fi
+
+    echo "Using OVMF_CODE: ${OVMF_CODE}"
+    echo "Using OVMF_VARS: ${OVMF_VARS}"
+
+    # generate nvram file using libvirt
+    virt-install --name constell-nvram-gen \
+        --description 'Constellation' \
+        --ram 1024 \
+        --vcpus 1 \
+        --osinfo detect=on,require=off \
+        --disk "${image_path},format=raw" \
+        --boot "machine=q35,menu=on,loader=/usr/share/OVMF/OVMF_CODE_4M.ms.fd,loader.readonly=yes,loader.type=pflash,nvram.template=/usr/share/OVMF/OVMF_VARS_4M.ms.fd,nvram=${BASE_DIR}/image.nvram,loader_secure=yes" \
+        --features smm.state=on \
+        --noautoconsole
+    echo 'connect using'
+    echo 'virsh console constell-nvram-gen'
+    echo ''
+    echo 'either load db cert using mokmanager or enroll full PKI using firmware setup'
+    echo ''
+    echo '    For mokmanager, try to boot as usual. You will see this message:'
+    echo '    > "Verification failed: (0x1A) Security Violation"'
+    echo '    Press OK, then ENTER, then "Enroll key from disk"'
+    echo '    Select the following key:'
+    echo '    > /EFI/loader/keys/auto/db.cer'
+    echo '    Press Continue, then choose "Yes" to the question "Enroll the key(s)?"'
+    echo '    Choose reboot and continue this script.'
+    echo ''
+    echo '    For firmware setup, press F2.'
+    echo '    Go to "Device Manager">"Secure Boot Configuration">"Secure Boot Mode"'
+    echo '    Choose "Custom Mode"'
+    echo '    Go to "Custom Securee Boot Options"'
+    echo '    Go to "PK Options">"Enroll PK", Press "Y" if queried, "Enroll PK using File"'
+    echo '    Select the following cert: /EFI/loader/keys/auto/PK.cer'
+    echo '    Choose "Commit Changes and Exit"'
+    echo '    Go to "KEK Options">"Enroll KEK", Press "Y" if queried, "Enroll KEK using File"'
+    echo '    Select the following cert: /EFI/loader/keys/auto/KEK.cer'
+    echo '    Choose "Commit Changes and Exit"'
+    echo '    Go to "DB Options">"Enroll Signature">"Enroll Signature using File"'
+    echo '    Select the following cert: /EFI/loader/keys/auto/db.cer'
+    echo '    Choose "Commit Changes and Exit"'
+    echo '    Repeat the last step for the following certs:'
+    echo '    > /EFI/loader/keys/auto/MicWinProPCA2011_2011-10-19.crt'
+    echo '    > /EFI/loader/keys/auto/MicCorUEFCA2011_2011-06-27.crt'
+    echo '    Reboot and continue this script.'
+    echo ''
+    echo 'Press ENTER to continue after you followed one of the guides from above.'
```

Comment:
I don't have the visual output in front of me, but I can imagine this is a wall of text that requires careful reading.

I think this could be a great candidate for some ANSI escape code formatting for highlighting file names and differentiating between "mokmanager" and "Enroll full PKI" steps.